### PR TITLE
Allow Setting Headers Per Request

### PR DESCRIPTION
**Pull request by s0kil**
Closes https://github.com/digitallyinduced/ihp/issues/273

### CODE REVIEW

**Comment by s0kil** [comment]
Should add some documentation, and to the Guide
@mpscholten Any ideas how to improve this implementation?

**Comment by s0kil** [comment]
Could probably also add tests for this

**Comment by s0kil** [comment]
Good to go

File: Guide/controller.markdown [hunk 1]
```diff
@@ -266,6 +266,13 @@ to `Just "the user agent value"`.
 
 The lookup for the header in the request is case insensitive.
 
+Use `setHeader` to set a request header:
+
+```haskell
+action ExampleAction = do
+    setHeader ("Max-Age", 10)
```
Comment (mpscholten) [inline Guide/controller.markdown:273]:
```suggestion
    setHeader ("Max-Age", "10")
```

File: IHP/ControllerSupport.hs [hunk 2]
```diff
@@ -169,6 +172,40 @@ getHeader :: (?context :: ControllerContext) => ByteString -> Maybe ByteString
...
+setHeader header = do
+    maybeHeaders <- Context.maybeFromContext @[Header]
+    let headers = fromMaybe [] maybeHeaders
+    Context.putContext (header : headers)
+{-# INLINABLE setHeader #-}
+
+-- | Add headers to current response
+-- | Returns a Response with headers
+--
+-- >>>  addResponseHeaders [("Content-Type", "text/html")] response
```
Comment (mpscholten) [inline IHP/ControllerSupport.hs:189]:
```suggestion
-- > addResponseHeaders [("Content-Type", "text/html")] response
```

The `>>>` haddock syntax always expects a result object in the next line. As it's not here we need to use a `>`

File: IHP/ControllerSupport.hs [hunk 3]
```diff
@@ -169,6 +172,40 @@ getHeader :: (?context :: ControllerContext) => ByteString -> Maybe ByteString
...
+--
+addResponseHeaders :: [Header] -> Response -> Response
+addResponseHeaders headers = Network.Wai.mapResponseHeaders (\hs -> headers ++ hs)
+{-# INLINABLE addResponseHeaders #-}
+
+-- | Add headers to current response, getting the headers from ControllerContext
+-- | Returns a Response with headers
+--
+-- >>>  addResponseHeadersFromContext response
```
Comment (mpscholten) [inline IHP/ControllerSupport.hs:198]:
```suggestion
-- > addResponseHeadersFromContext response
```

File: IHP/ControllerSupport.hs [hunk 4]
```diff
@@ -169,6 +172,40 @@ getHeader :: (?context :: ControllerContext) => ByteString -> Maybe ByteString
...
+    Context.putContext (header : headers)
+{-# INLINABLE setHeader #-}
+
+-- | Add headers to current response
+-- | Returns a Response with headers
+--
+-- >>>  addResponseHeaders [("Content-Type", "text/html")] response
+--
+addResponseHeaders :: [Header] -> Response -> Response
+addResponseHeaders headers = Network.Wai.mapResponseHeaders (\hs -> headers ++ hs)
```
Comment (mpscholten) [inline IHP/ControllerSupport.hs:192]:
```suggestion
addResponseHeaders headers = Network.Wai.mapResponseHeaders (\hs -> headers <> hs)
```

We use `<>` in IHP generally (`++` is aliased to `<>`)

**Review by mpscholten — COMMENTED** [review]
Looks good 👍 I think the documentation is fine

**Review by mpscholten — APPROVED** [review]
Thanks 👍

**Pull request merged**